htmresearch/frameworks/clustering/sdr_clustering.py

```diff
@@ -0,0 +1,213 @@
+    self._label = label
+
+
+  def getLabel(self):
+    return self._label
+
+
+  def setValue(self, value):
+    self._value = value
+
+
+  def getValue(self):
+    return self._value
+
+
+
+class Cluster(object):
+  def __init__(self, id, points=None, label=None):
+    self._id = id
+    self._label = label
+    if points:
+      self._points = points
+    else:
+      self._points = []
+    _LOGGER.debug("New cluster %s created!" % self._id)
+
+
+  def add(self, point):
+    self._points.append(point)
+
+
+  def setLabel(self, name):
+    self._label = name
+
+
+  def getLabel(self):
+    return self._label
+
+
+  def getId(self):
+    return self._id
+
+
+  def setPoints(self, points):
+    self._points = points
+
+
+  def getPoints(self):
+    return self._points
+
+
+  def size(self):
+    return len(self._points)
+
+
+  def merge(self, cluster):
+    for point in cluster.getPoints():
+      self.add(point)
+
+
+
+class Clustering(object):
+  def __init__(self,
+               mergeThreshold,
+               anomalousThreshold,
+               stableThreshold,
+               fistClusterId=0):
+
+    # Clusters
+    self.newCluster = Cluster(fistClusterId)
+    self.clusters = []
+
+    # Anomaly Score Thresholds
+    self.anomalousThreshold = anomalousThreshold  # to create new cluster
+    self.stableThreshold = stableThreshold  # to add point to cluster
+
+    # Cluster distance threshold
+    self.mergeThreshold = mergeThreshold
+
+
+  def findClosestClusters(self, cluster):
+    """
+    
+    :param cluster: 
+    :return: ordered list of clusters and their distances 
+    """
+    dists = []
+    for c in self.clusters:
+      d = clusterDist(c.getPoints(), cluster.getPoints())
+      dists.append((d, c))
+
+    return sorted(dists)
+
+
+  def addCluster(self, cluster):
+    self.clusters.append(cluster)
+    _LOGGER.debug("Cluster %s permanently added" % cluster.getId())
+
+
+  def addOrMergeCluster(self, cluster):
+
+    clusterDistPairs = self.findClosestClusters(cluster)
+    if len(clusterDistPairs) > 0:
+      notMerged = True
+      for clusterDistPair in clusterDistPairs: 
+        closestClusterToNewDist, closestClusterToNew = clusterDistPair
+  
+        if closestClusterToNewDist < self.mergeThreshold:
+          _LOGGER.debug("Cluster %s merged with cluster %s. Inter-cluster "
+                        "distance: %s" % (cluster.getId(),
+                                          closestClusterToNew.getId(),
+                                          closestClusterToNewDist))
+          closestClusterToNew.merge(cluster)
+          notMerged = False
+      if notMerged:
+        self.addCluster(cluster)
+    else:
+      self.addCluster(cluster)
+
+
+  def infer(self):
+    """
+    Inference: find the closest cluster to the new cluster.
+    """
+    clusterDistPairs = self.findClosestClusters(self.newCluster)
+    if len(clusterDistPairs) > 0:
+      distToCluster, predictedCluster = clusterDistPairs[0]
```

Comment:
I don't think this scenario can happen. If the newCluster has 0 (or very low) distance to more than one clusters, doesn't this mean that the distance between the two clusters is 0? therefore one of these clusters should have been merged into the other one instead of being added as a new cluster.